power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -58,29 +28,115 @@ std::tuple<Idx, Idx> count_voltage_sensors(const Idx n_bus, const MeasuredValues
+        for (Idx ybus_index = y_bus_structure.bus_entry[row] + 1; ybus_index != y_bus_structure.row_indptr[row + 1];
+             ++ybus_index) {
+            for (Idx element_index = y_bus_structure.y_bus_entry_indptr[ybus_index];
+                 element_index != y_bus_structure.y_bus_entry_indptr[ybus_index + 1]; ++element_index) {
+                YBusElement const& element = y_bus_structure.y_bus_element[element_index];
+                // shunt should not be considered
+                // if the branch is fully connected and measured, we consider it as a valid flow sensor
+                // we only need one flow sensor, so the loop will break
+                if (element.element_type != YBusElementType::shunt) {
+                    Idx const branch = element.idx;
+                    if ((measured_values.has_branch_from(branch) || measured_values.has_branch_to(branch)) &&
+                        topo.branch_bus_idx[branch][0] != -1 && topo.branch_bus_idx[branch][1] != -1) {
+                        flow_sensors[ybus_index] = 1;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    return flow_sensors;
+}
+
+// re-organize the flow sensor for radial grid without phasor measurement
+// this mutate the flow sensor vector to try to assign injection sensor to branch sensor
+// all the branch should be measured if the system is observable
+// this is a sufficient condition check
+// if the grid is not radial, the behavior is undefined.
+inline void assign_injection_sensor_radial(YBusStructure const& y_bus_structure, std::vector<int8_t>& flow_sensors) {
+    Idx const n_bus = std::ssize(y_bus_structure.row_indptr) - 1;
+    // loop the row without the last bus
+    for (Idx row = 0; row != n_bus - 1; ++row) {
+        Idx const current_bus = row;
+        Idx const bus_entry_current = y_bus_structure.bus_entry[current_bus];
+        Idx const branch_entry_upstream = bus_entry_current + 1;
+        // there should be only one upstream branch in the upper diagonal
+        // so the next of branch_entry_upstream is already the next row
+        // because the grid is radial
+        assert(y_bus_structure.row_indptr[current_bus + 1] == branch_entry_upstream + 1);
+        Idx const upstream_bus = y_bus_structure.col_indices[branch_entry_upstream];
+        Idx const bus_entry_upstream = y_bus_structure.bus_entry[upstream_bus];
+        // if the upstream branch is not measured
+        if (flow_sensors[branch_entry_upstream] == 0) {
+            if (flow_sensors[bus_entry_current] == 1) {
+                // try to steal from current bus
+                flow_sensors[bus_entry_current] = 0;
+                flow_sensors[branch_entry_upstream] = 1;
+            } else if (flow_sensors[bus_entry_upstream] == 1) {
+                // if not possible, steal from upstream bus
+                flow_sensors[bus_entry_upstream] = 0;
+                flow_sensors[branch_entry_upstream] = 1;
```

Comment:
does it make sense to use `swap`?
```suggestion
                // try to steal from current bus
                std::swap(flow_sensors[branch_entry_upstream], flow_sensors[bus_entry_current]);
            } else if (flow_sensors[bus_entry_upstream] == 1) {
                // if not possible, steal from upstream bus
                std::swap(flow_sensors[branch_entry_upstream], flow_sensors[bus_entry_upstream]);
```